bpo-24024: update str.__doc__ to show default encoding explicitly

plusminushalf opened the pull request:
sys.getdefaultencoding() always returns 'utf-8'
Reference: Objects/unicodeobject.c line number -> 4214

Also str() accepts bytes, bytearray or buffer-like objects as input
hence changing it to object.
As str returns .\_\_str\_\_() method defined inside object or repr(obj)
otherwise.

plusminushalf committed:
bpo-24024 update str.__doc__ to show default encoding explicitly

sys.getdefaultencoding() always returns 'utf-8'
Reference: Objects/unicodeobject.c line number -> 4214

Also str() accepts bytes, bytearray or buffer-like objects as input
hence changing it to object.
As str returns .__str__() method defined inside object or repr(obj)
otherwise.

plusminushalf commented:
```python
>>> str.__doc__
'str(object=\'\') -> str\nstr(object, encoding="utf-8", errors="strict") -> str\n\nCreate a new string object from the given object. If encoding or\nerrors is specified, then the object must expose a data buffer\nthat will be decoded using the given encoding and error handler.\nOtherwise, returns the result of object.__str__() (if defined)\nor repr(object).\nencoding defaults to sys.getdefaultencoding().\nerrors defaults to \'strict\'.'
```

vadmium commented on Objects/unicodeobject.c:
```diff
@@ -15120,7 +15120,7 @@ unicode_subtype_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
 PyDoc_STRVAR(unicode_doc,
 "str(object='') -> str\n\
-str(bytes_or_buffer[, encoding[, errors]]) -> str\n\
+str(object, encoding=\"utf-8\", errors=\"strict\") -> str\n\
```
Best to use single quotes to be consistent with the line above.

Since you add the defaults to the signature, I think you can remove them from the end of the text below.

vadmium reviewed (commented):
In general this looks worthwhile to me, just some minor suggestions

plusminushalf committed:
Code review changes

1. Changing double quotes to single
2. Removing defaults information from the end since it is already
defined in function signature

DimitrisJim commented on Objects/unicodeobject.c:
```diff
@@ -15120,15 +15120,13 @@ unicode_subtype_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
 PyDoc_STRVAR(unicode_doc,
 "str(object='') -> str\n\
-str(bytes_or_buffer[, encoding[, errors]]) -> str\n\
+str(object, encoding='utf-8', errors='strict') -> str\n\
```
I'm not seeing why this needs two signatures now when `str(object='', encoding='utf-8', errors='strict')` would suffice.

merwok replied:
I don’t think that merging the two lines is correct.

plusminushalf committed:
Removing extra function signatures

vadmium commented on Objects/unicodeobject.c:
```diff
@@ -15119,8 +15119,7 @@ unicode_subtype_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
 PyDoc_STRVAR(unicode_doc,
-"str(object='') -> str\n\
-str(object, encoding='utf-8', errors='strict') -> str\n\
+"str(object, encoding='utf-8', errors='strict') -> str\n\
```
Passing no arguments is also supported:

```python console
>>> str()
''
```

DimitrisJim replied:
Indeed, object needs a default value, i.e `object=''`.

plusminushalf committed:
Code review changes

serhiy-storchaka commented on Objects/unicodeobject.c:
```diff
@@ -15119,16 +15119,13 @@ unicode_subtype_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
 PyDoc_STRVAR(unicode_doc,
-"str(object='') -> str\n\
-str(bytes_or_buffer[, encoding[, errors]]) -> str\n\
+"str(object='', encoding='utf-8', errors='strict') -> str\n\
```
There is a problem with this. This signature means that `str('', 'utf-8', 'strict')` is equivalent to `str()`. But calling `str` with string argument and encoding is an error. The signature of `str` can't be expressed in such form, this is a cause why the `str` constructor still is not converted to Argument Clinic.

merwok replied:
I don’t think that merging the two lines is correct.

serhiy-storchaka reviewed (changes requested):
Sorry, it was approved by mistake. I don't approve this patch in the current form.
